File: Cisco/7HouseRobber.py
class Solution:
    def rob(self, nums: list[int]) -> int:
        N = len(nums)

        if N == 1:
            return nums[0]
        if N == 2:
            return max(nums)

        prevprev = 0
        prev = nums[-1]
        cur = nums[-2]

        for i in range(N - 3, -1, -1):
            cur, prev, prevprev = nums[i] + max(prev, prevprev), cur, prev

        return max(cur, prev)

File: Cisco/GitHubList/C++/7HouseRobber.cpp
#include <vector>

using namespace std;

class Solution
{
public:
    int rob(vector<int> &nums)
    {
        int N = nums.size();
        if (N == 1)
            return nums[0];
        if (N == 2)
            return max(nums[0], nums[1]);

        int prev = nums[N - 2],
            prevprev = nums[N - 1],
            prevprevprev = 0;
        int cur;

        for (int i = N - 3; i >= 0; --i)
            cur = nums[i] + max(prevprev, prevprevprev),
            prevprevprev = prevprev,
            prevprev = prev,
            prev = cur;

        return max(prev, prevprev);
    }
};